source/opt/instruction.cpp

```diff
@@ -133,5 +142,186 @@ void Instruction::ReplaceOperands(const std::vector<Operand>& new_operands) {
   operands_.insert(operands_.begin(), new_operands.begin(), new_operands.end());
   operands_.shrink_to_fit();
 }
+
+bool Instruction::IsReadOnlyLoad() const {
+if (IsLoad()) {
+  ir::Instruction* address_def = GetBaseAddress();
+  if (address_def->opcode() == SpvOpLoad ||
+      address_def->opcode() == SpvOpFunctionParameter) {
+    return false;
+  }
+  assert(address_def->opcode() == SpvOpVariable &&
+         "Expecting to get a variable.");
+
+  return address_def->IsReadOnlyVariable();
+}
+return false;
+}
+
+Instruction* Instruction::GetBaseAddress() const {
+assert(IsLoad() &&
+       "GetBaseAddress should only be called on load instructions.");
+uint32_t base = GetSingleWordInOperand(kLoadBaseIndex);
+ir::Instruction* base_inst = context()->get_def_use_mgr()->GetDef(base);
+bool done = false;
+while (!done) {
+  switch (base_inst->opcode()) {
+    case SpvOpAccessChain:
+    case SpvOpInBoundsAccessChain:
+    case SpvOpPtrAccessChain:
+    case SpvOpInBoundsPtrAccessChain:
+    case SpvOpImageTexelPointer:
+      base = base_inst->GetSingleWordInOperand(kLoadBaseIndex);
```

Comment:
Fixed the first two comments.  I'll leave the third.  In that case we will return the bitcast as the base pointer, and value numbering will assume it is not read only.  If there is a specific case others want to get it can be extended to get look through those instructions as well.